neuralop/layers/coda_blocks.py

```diff
@@ -0,0 +1,380 @@
+        rank (float): Rank of the factorization, by default 1.0.
+        spectral_convolution (callable): Spectral convolution module to be used.
+        joint_factorization (bool): whether to factorize all spectralConv weights as one tensor, by default False
+
+        # Normalization
+        Normalizer (callable): Normalization module to be used.
+
+        codimension_size (int): Size of the codimension for whole function, only used for permutation_eq = False.
+        per_channel_attention (bool): whether to use per channel attention, by default True (overwrite token_codimension to 1).
+        permutation_eq (bool): whether to use permutation equivariant mixer layer  after attention mechanism.
+        temperature (float): Temperature parameter for attention mechanism.
+        nonlinear_attention (bool): whether to use non-linear activation for K,Q,V operator.
+        scale (int): Scale for downsampling Q,K functions before calculating the attention matrix. Higher scale will downsample more.
+    """
+
+    def __init__(
+        self,
+        n_modes,
+        n_heads=1,
+        token_codimension=1,
+        output_scaling_factor=None,
+        incremental_n_modes=None,
+        head_codimension=None,
+        use_channel_mlp=False,
+        non_linearity=F.gelu,
+        preactivation=False,
+        fno_skip='linear',
+        channel_mlp_skip='linear',
+        channel_mlp_expansion=1.0,
+        separable=False,
+        factorization='tucker',
+        rank=1.0,
+        spectral_convolution=None,
+        norm="instance_norm",
+        joint_factorization=False,
+        fixed_rank_modes=False,
+        implementation='factorized',
+        decomposition_kwargs=None,
+        fft_norm='forward',
+        codimension_size=None,
+        per_channel_attention=True,
+        permutation_eq=True,
+        temperature=1.0,
+        nonlinear_attention=False,
+        scale=None,
+        **_kwargs,
+    ):
+        super().__init__()
+
+        # Co-dimension of each variable/token. The token embedding space is
+        # identical to the variable space, so their dimensionalities are equal.
+        self.token_codimension = token_codimension
+
+        # codim of attention from each head
+        self.head_codimension = (head_codimension
+                                 if head_codimension is not None
+                                 else token_codimension)
+        self.n_heads = n_heads  # number of heads
+        self.output_scaling_factor = output_scaling_factor
+        self.temperature = temperature
+        self.num_dims = len(n_modes)
+
+        if norm is None:
+            Normalizer = torch.nn.Identity
+        elif norm == "instance_norm":
+            Normalizer = partial(nn.InstanceNorm2d, affine=True) if self.num_dims == 2 else partial(
+                nn.InstanceNorm3d, affine=True)
+        else:
+            raise ValueError(f"Unknown normalization type {norm}")
+
+        if spectral_convolution is None:
```

Comment:
Have you tested this layer with `SphericalConv` blocks? Good to keep this option if so, but if not, better to fix the conv module.

Reply:
`spectral_convolution` is passed as argument. So if required, one can pass `SphericalConv`.
What additional 'option' are you suggesting?